Returning a per request metric for number of cached_tokens read

havetc opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

Right now sglang uses an advanced radix cache system, but it is not possible to know for each request how many of the
tokens were computed, or read from the cache. This pull request offer a way to easily return this information

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications

Most important edit is to the UsageInfo structure, adding `cached_tokens` in `prompt_tokens_details` the same way [openai](https://openai.com/index/api-prompt-caching/) does it. To limit the impact to current users, this is disabled by default and needs to be activated using `--activate-cache-report` server parameters.
For sglang `/generate` requests, the number of cached tokens is always returned in the `meta_info` dictionary


I'm not really sure about the extra parameter to the server, which is not implemented very elegantly. I'm open to suggestion to improve it. It could also be entirely removed and this report be activated by default (to avoid adding yet another server argument).
<!-- Describe the changes made in this PR. -->

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

havetc committed:
[WIP] Implementation of a cache counter in sglang

we want to know when we get the result how much of it was computed
vs retrieved from the cache

havetc committed:
[WIP] Cache measure

kinda works ATM. Need to test chunked prefill, what happens if the
prefill is cut because too much tokens are present to be done in one
pass
also need to test concurrent request

havetc committed:
[DEV] Changing formula, adding tests

tests of chunked prefill showed that the cache computation doesn't
wrongly count token of previous chunks as already in the cache, so
we removed the - len(self.prefix_indices) accordingly

new test allow for the test of chunked prefill, and also another test
check how it works when two concurrent request are executed in the
same batch

havetc committed:
[DEV] Removing local parameter

allowing for tests on my rtx2070, but not really consistent with the
rest of the tests

havetc committed:
[DEV] Refacto with pre-commit

havetc committed:
[DEV] Changing to openai API, +make it activable

Adding --activate-cache-report server arg to activate cache report

havetc committed:
[FIX] removing wrong debug print

havetc committed:
[FIX] local changes for tests were committed by mistake

merrymercy commented on python/sglang/srt/server_args.py:
```diff
@@ -72,6 +72,7 @@ class ServerArgs:
     # Other
     api_key: Optional[str] = None
     file_storage_pth: str = "SGLang_storage"
+    activate_cache_report: bool = False
```
to make it more consist with other names, rename it to `enable_cache_report`

merrymercy commented on test/srt/test_cache_report.py:
```diff
@@ -0,0 +1,185 @@
+            timeout=300,
+            other_args=[
+                "--chunked-prefill-size=40",
+                "--disable-cuda-graph",
```
why do you disable cuda graph?

havetc replied:
I can remove it for better consistency with other tests. I run my test locally on an rtx2070m, so I need to disable cuda graph (and also change the model to gemma 2b) for the test to not go out of memory.

merrymercy reviewed (commented):
This is a great feature to have! I left some comments. Please fix them.

havetc committed:
[DEV] Fixing first PR comments

changing arg name to enable-cache-report
changing test args (remove disable cuda graphs)
and another addition
-fix missing call

havetc committed:
[DEV] Fixing wrong cache computation

it was not computed at the right place
also, tests were updated

havetc commented:
I changed the computation as suggested. However, it was not that simple, as it now requires to explicitely handle the case where a prefill is split among several chunks (to avoid counting the first chunks as being already in cache).
It seems to work at the moment, but I'll test it more to be really sure (probably on Monday).

I'll also try to handle the case where the same prompt was already sent, it seems in that case that the cache is prompt_tokens - 1, which isn't really true it should be prompt_tokens.

merrymercy commented:
The case of sending the same prompt twice is tricky. It depends on how you define "cache tokens".
1. If you treat it as cached tokens in memory, then the second time, #cache-tokens should be #prompt-token
2. If you define #cached-tokens as the number of skipped tokens in the prefill, then the current return value #prompt-token - 1 is correct. Because even if you cache the full prompt, you only cache the KV cache, not the final logits. You have to run the computation again for that last token to get the logits.

I think returning #prompt-token - 1 for the second time makes sense from a computation-saving perspective.

havetc commented:
@merrymercy I did some further tests, and it seems that now it counts properly, I didn't saw any request with the cache counted several time. I let the return be #prompt-token -1 as it makes more sense, and it doesn't make a big difference anyway.

I rebased the branch, it should be "almost" up to date (a commit on the main branch occured at the same time it seems)

havetc commented:
@merrymercy chunked prefill is already tested and even present in the unit test (that's why the engine is configured with `--chunked-prefill-size=40` in this test, to test a prefill splitted between several chunks without it being to big / long to process).
I just tested jump-forward decoding, but it seems I'm affected by this bug #1312 and I'm not sure that it only impacts the logs.
With and without jump forward activated, for a simple schema with an enum (where in theory jump forward should be able to jump the length of the enum, that I made extra long for the test), I got the exact same Usage in output (prompt, completion and cached), and even printing the internal `completion_tokens_wo_jump_forward` they are the same . I'm not sure that's normal.

merrymercy commented:
@havetc Thanks for the contribution. It is merged. We can iterate on this and fix bugs if there is any mismatch on the token counting.

yudian0504 commented on python/sglang/srt/managers/schedule_batch.py:
```diff
@@ -496,6 +499,13 @@ def prepare_for_extend(self, vocab_size: int):
+                if req.extend_logprob_start_len > 0
+                else 0
+            )
+            req.cached_tokens += len(req.prefix_indices) - already_computed
```
the cached_tokens got an  incorrect value when enably chunked prefill. seems incompatible?
